File: Prefix Suffix String.cpp
class TrieNode
{
    public: 
    char data ;
    TrieNode* children[26] ;
    bool ister; 
    TrieNode(char ch )
    {
        data =ch; 
        for(int i = 0 ;i<26; i++)
        {
            children[i] = NULL ;
        }
        ister = false;
    }
    
    
};

class Solution{
public:
      TrieNode* root = new TrieNode('0') ;
      
      void insertutil(TrieNode*  root , string  word )
      {
          
          if(word.length()==0)
          {
              root->ister = true; 
              return ; 
          }
          
          TrieNode* child; 
          
          //present
          if(root->children[word[0]-'a'])
          {
              child = root->children[word[0]-'a']  ;
          }
          else // absent
          {
              child = new TrieNode(word[0]);
              root->children[word[0]-'a'] = child ; 
          }
           
          insertutil(child , word.substr(1)) ;
          
      }
      
      bool cal(TrieNode* root , string word , int ind  )
      {
          
          if(ind==word.length())
          {
              return true ; 
          }
          
          
          bool ans = false ;
          
        
                  if(root->children[word[ind]-'a']==NULL)
                  {
                      return false ;
                   
                  }
                  else
                  {
                      ans|=cal(root->children[word[ind]-'a'] ,word, ind+1  ) ;
                  }
        return ans ;   
      }
      
      
      
      
      
    int prefixSuffixString(vector<string> s1,vector<string> s2){
        // Code here
        
        for(auto &  it: s1)
         {
             insertutil(root , it);
             reverse(begin(it), end(it)) ; 
             insertutil(root , it);
             
         }
         
         int ans = 0 ;
         
         for(auto & it : s2 )
         {
             string s = it ;
             reverse(begin(s), end(s)) ;
             if(cal(root , it ,0  ) || cal(root, s, 0))
             ans++ ;
            //  cout << ans << endl ;
             
         }
         return ans ; 
    }
};

File: Prefix Suffix String.java

class TrieNode{
    boolean isEnd;
    TrieNode children[];
    TrieNode(){
        children=new TrieNode[26];
        Arrays.fill(children,null);
        isEnd=false;
    }
}
class Solution
{
    
    public void insert(TrieNode root,String str)
    {
        TrieNode child=root;
        for(int i=0;i<str.length();i++)
        {
            char ch=str.charAt(i);
            if(child.children[ch-'a']==null){
                child.children[ch-'a']=new TrieNode();
            }
            child=child.children[ch-'a'];
        }
    }
    public boolean find(TrieNode root,String s2)
    {
        TrieNode child=root;
        for(int i=0;i<s2.length();i++)
        {
            char ch=s2.charAt(i);
            if(child.children[ch-'a']==null){
                return false;
            }
            child=child.children[ch-'a'];
        }
        return true;
    }
    
    public int prefixSuffixString(String s1[],String s2[])
    {
        TrieNode root=new TrieNode();
        for(int i=0;i<s1.length;i++){
            insert(root,s1[i]);
            StringBuilder str=new StringBuilder(s1[i]);
            String reverseString=str.reverse().toString();
            insert(root,reverseString);
        }
        int count=0;
        for(int i=0;i<s2.length;i++){
            StringBuilder str=new StringBuilder(s2[i]);
            String revStr=str.reverse().toString();
            if(find(root,s2[i])||find(root,revStr)){
                count++;
            }
        }
        return count;
    }
}
